### PR TITLE
[RPC] test diagnostics handling

### PR DESCRIPTION
**Pull request by rgrinberg**
Given the upcoming changes to this stuff, it's better to make sure we
have some tests to verify nothing breaks. I did not write this in cram
because I expect to test other client functionality that will not be
available via just cli commands.

### CODE REVIEW

File: test/expect-tests/dune_rpc_e2e/dune_rpc_e2e.ml [hunk 1]
```diff
@@ -0,0 +1,401 @@
...
+
+let read_lines in_ =
+  let* reader = Scheduler.Worker.create () in
+  let in_ = Unix.in_channel_of_descr in_ in
+  let rec loop acc =
+    let* res = Scheduler.Worker.task reader ~f:(fun () -> input_line in_) in
+    match res with
+    | Ok a -> loop (a :: acc)
+    | Error `Stopped -> assert false
+    | Error (`Exn _) -> Fiber.return (String.concat (List.rev acc) ~sep:"\n")
```
Comment (aalekseyev) [inline test/expect-tests/dune_rpc_e2e/dune_rpc_e2e.ml:54]: Print out the exn? (not sure if it's reachable or what, but hiding errors generally makes things harder to debug)

**Review by aalekseyev — APPROVED** [review]
It's great to have more automated tests of watching mode! Looks good.

**Comment by aalekseyev** [comment]
... except the tests seem to time out, which should be investigated.

**Comment by aalekseyev** [comment]
I looked a bit and found two things:
- We don't close stdout and stderr of the child process on the side of the test runner; I don't know if that's necessarily a problem given the workaround below, but it's definitely a bug. I pushed a fix.
- The same `read_packets` loop that was getting stuck before is getting stuck again. Why do you expect `let+ res = Client.connect ?handler chan initialize ~f in` on line 42 to ever terminate?

By the way, this is quite painful to debug. I ended up doing the following:
- replace `let%expect_test "turn on and shutdown" =` with `let () = `
- figure out how to run the command dune runs outside of dune
- add debug printfs and run the exe in my terminal (tweaking PATH by hand so it calls the right dune)

**Review by aalekseyev — CHANGES_REQUESTED** [review]
^

**Comment by rgrinberg** [comment]
> We don't close the write ends of stdout and stderr of the child process on the side of the test runner; I don't know if that's necessarily a problem given the workaround below, but it's definitely a bug. I pushed a fix.

Thanks. that makes sense. I wonder how it worked for me locally without this.

> The same read_packets loop that was getting stuck before is getting stuck again. Why do you expect let+ res = Client.connect ?handler chan initialize ~f in on line 42 to ever terminate?

Because eventually reading from `chan` will return eof once it channel is closed.

> By the way, this is quite painful to debug. I ended up doing the following:

One thing that helped me was to kill subprocesses when they got suck. Viewing the output was never a problem because I could use dtrace. To make this easier to debug, how abot we add a simple timeout mechanism to `Scheduler.wait_for_process`?

**Comment by aalekseyev** [comment]
> Because eventually reading from `chan` will return eof once it channel is closed.

And why will that happen? 
Is that supposed to be triggered by `let+ () = Chan.write chan None in` in `connect_raw`? (which flows to the dune process and then back to the client process?)
Any known issues with that?

> kill subprocesses when they got s[t]uck

Hmm, I think it didn't help me at first, but I can see that.

> Viewing the output was never a problem because I could use dtrace.

How do you use it? What does it let you see?

> how abo[u]t we add a simple timeout mechanism to Scheduler.wait_for_process?

Doesn't that come down to having a timer events in the scheduler, which was a bit problematic? (that said, I have nothing against it even if it's done inefficiently with one thread per timeout, tbh)

**Comment by rgrinberg** [comment]
> Is that supposed to be triggered by let+ () = Chan.write chan None in in connect_raw? (which flows to the dune process and then back to the client process?)

That's right. writing `None` will trigger:
```ocaml
  let close t =
    match t.state with
    | Closed -> ()
    | Open { in_channel; out_channel; reader; writer; socket } ->
      Worker.stop reader;
      Worker.stop writer;
      (* with a socket, there's only one fd. We make sure to close it only once.
         with dune rpc init, we have two separate fd's (stdin/stdout) so we must
         close both. *)
      if not socket then close_in_noerr in_channel;
      close_out_noerr out_channel;
      t.state <- Closed
```

> Any known issues with that?

I suppose if `f client` raises then it will not close. I could definitely improve this bit. However, I'm currently in the process of getting rid of `dune rpc init` entirely. So I'd rather wait for that first.

> How do you use it? What does it let you see?

I get to see everything read/written by the test and its subprocesses. I just remembered that dtrace isn't a thing on Linux. That'a a shame, but i'm pretty sure something equivalent can be done with strace. For reference, this is what works for me:

```
#!/usr/sbin/dtrace -qs
#pragma D option strsize=10000
inline int COPYINSTRLIMIT = 128;

BEGIN {
  test_pid = -1;
}

syscall::write*:entry
/pid == test_pid || ppid == test_pid/
{
  printf("%s(%d) > %s\n", execname, pid, copyinstr(arg1));
}

syscall::read*:entry
/pid == test_pid || ppid == test_pid/
{
  printf("%s(%d) < %s\n", execname, pid, copyinstr(arg1));
}

proc:::create /execname == "inline_test_runner_dune_rpc_e2e."/ {
  test_pid = pid;
}
```

> Doesn't that come down to having a timer events in the scheduler, which was a bit problematic?

Why is it problematic?

> (that said, I have nothing against it even if it's done inefficiently with one thread per timeout, tbh)

I planned on having one thread with a polling loop.

**Comment by aalekseyev** [comment]
> I suppose if `f client` raises then it will not close.

That's not what happened, I think. Pretty sure I saw `Chan.write chan None` happen.

> this is what works for me:

Oh, that's some involved config. I don't know if strace configs are as expressive. I could definitely get some of that info with strace, good idea.

> Why is it problematic?

Not so much problematic, only a feature we don't yet have.

**Comment by rgrinberg** [comment]
> Not so much problematic, only a feature we don't yet have.

Added a timeout mechanism [here](https://github.com/ocaml/dune/pull/4751). Once it is reviewed and merged, I will update this PR.

File: test/expect-tests/dune_rpc_e2e/dune_rpc_e2e.ml (hunk 1 above)
Comment (rgrinberg) [reply test/expect-tests/dune_rpc_e2e/dune_rpc_e2e.ml]: Done.

**Comment by rgrinberg** [comment]
And now I updated the PR to use rpc to shutdown/build

**Comment by rgrinberg** [comment]
@aalekseyev another round of review would be appreciated. I will add timeouts once #4751 is merged.

**Comment by rgrinberg** [comment]
@aalekseyev this should be ready.  I changed the code use direct connect (and fixed some bugs along the way), and I can't make it get it to hang anymore.

**Comment by aalekseyev** [comment]
Doesn't hang for me, either.

I'm seeing a diff in test, though:

```
-    Warning 26 [unused-var]: unused variable x.
+   Warning 26: unused variable x.
```

Is this something weird in your setup, in my setup, or "nobody's fault"?

**Comment by aalekseyev** [comment]
I have OCaml 4.11.1 by the way.

**Comment by rgrinberg** [comment]
> I have OCaml 4.11.1 by the way.

That's the issue then. The error format changed a little in 4.12.

File: otherlibs/dune-rpc-lwt/src/dune_rpc_lwt.ml [hunk 2]
```diff
@@ -92,15 +92,20 @@ module V1 = struct
 
         let read_file s = Lwt_io.with_file ~mode:Input s Lwt_io.read
 
-        let readlink s = Lwt_unix.readlink s
+        let readlink s =
```
Comment (aalekseyev) [inline otherlibs/dune-rpc-lwt/src/dune_rpc_lwt.ml:95]: I'm not sure how general this function is supposed to be, but there's a race here between `readlink` and somebody modifying the directory.
Probably not a problem, but if I had a choice between having the race in a shared library or at the call site I'd probably choose at the call site. (feel free to ignore if you disagree)

Comment (rgrinberg) [reply otherlibs/dune-rpc-lwt/src/dune_rpc_lwt.ml:95]: How about an implementation that is race free but relies on exceptions:

```ocaml
try
  Some (Unix.readlink s)
with Unix.Unix_error (EINVAL, _, _) -> None
```

File: otherlibs/dune-rpc/private/dune_rpc_private.ml [hunk 3]
```diff
@@ -87,28 +87,37 @@ module Where = struct
...
-              let* link = Sys.readlink file in
-              loop link (symlink_depth - 1)
-            else
-              of_file file
-          | `Other -> Fiber.return None
-          | `Unix_socket -> Fiber.return (Some (`Unix file))
-          | `Normal_file -> of_file file
-        in
-        loop file 5
+        let* analyze = Sys.analyze_path file in
```
Comment (aalekseyev) [inline otherlibs/dune-rpc/private/dune_rpc_private.ml:96]: I guess there are slightly bizarre things going on here (both before and after your change), but why are you changing this code, anyway?

File: otherlibs/dune-rpc-lwt/src/dune_rpc_lwt.ml (hunk 2 above)
Comment (aalekseyev) [reply otherlibs/dune-rpc-lwt/src/dune_rpc_lwt.ml:95]: I like that better, yeah. (I'm assuming here `Unix.readlink` is the blocking one, not lwt one?)

File: otherlibs/dune-rpc/private/dune_rpc_private.ml (hunk 3 above)
Comment (rgrinberg) [reply otherlibs/dune-rpc/private/dune_rpc_private.ml:96]: It was completely broken before. The point of all these hacks is to get around the 104/108 (bsd/linux) limit on unix sockets. If the path is too long, we can use resolve and pass the symlink instead. The actual socket could be created elsewhere where the path is shorter.

File: otherlibs/dune-rpc-lwt/src/dune_rpc_lwt.ml (hunk 2 above)
Comment (rgrinberg) [reply otherlibs/dune-rpc-lwt/src/dune_rpc_lwt.ml:95]: Yes. For lwt, I will `Lwt_unix.readlink` and `Lwt.catch` respectively.

Comment (rgrinberg) [reply otherlibs/dune-rpc-lwt/src/dune_rpc_lwt.ml:95]: Switched to this implementation

**Comment by rgrinberg** [comment]
@aalekseyev the races are gone. One last round of review please

**Review by aalekseyev — APPROVED** [review]
Sorry, meant to approve on Thursday, forgot to click the button.

**Pull request merged**